### ISSUE TITLE
[Bug]: Unable to download Librosa on python 3.13.0 due to Numba support issue 

### ISSUE THREAD

**Issue by cdbruyla** [librosa/librosa#1883]
### Bug summary

Numba does not support 3.13.0 as of yet. Given Librosa depends on Numba, Librosa can't be installed on python 3.13.0 (only tried with pip). 
This issue is also being tracked in the Numba repo: https://github.com/numba/numba/issues/9413 

### Code for reproduction

```Python
pip install librosa
```

### Actual outcome

Defaulting to user installation because normal site-packages is not writeable
Collecting librosa
  Using cached librosa-0.10.2.post1-py3-none-any.whl.metadata (8.6 kB)
Collecting audioread>=2.1.9 (from librosa)
  Using cached audioread-3.0.1-py3-none-any.whl.metadata (8.4 kB)
Requirement already satisfied: numpy!=1.22.0,!=1.22.1,!=1.22.2,>=1.20.3 in ./.local/lib/python3.13/site-packages (from librosa) (2.1.2)
Requirement already satisfied: scipy>=1.2.0 in ./.local/lib/python3.13/site-packages (from librosa) (1.14.1)
Requirement already satisfied: scikit-learn>=0.20.0 in ./.local/lib/python3.13/site-packages (from librosa) (1.5.2)
Requirement already satisfied: joblib>=0.14 in ./.local/lib/python3.13/site-packages (from librosa) (1.4.2)
Requirement already satisfied: decorator>=4.3.0 in ./.local/lib/python3.13/site-packages (from librosa) (5.1.1)
Collecting numba>=0.51.0 (from librosa)
  Using cached numba-0.60.0.tar.gz (2.7 MB)
  Installing build dependencies ... done
  Getting requirements to build wheel ... error
  error: subprocess-exited-with-error
  
  × Getting requirements to build wheel did not run successfully.
  │ exit code: 1
  ╰─> [24 lines of output]
      Traceback (most recent call last):
        File "/usr/lib/python3.13/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 353, in <module>
          main()
          ~~~~^^
        File "/usr/lib/python3.13/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 335, in main
          json_out['return_val'] = hook(**hook_input['kwargs'])
                                   ~~~~^^^^^^^^^^^^^^^^^^^^^^^^
        File "/usr/lib/python3.13/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 118, in get_requires_for_build_wheel
          return hook(config_settings)
        File "/tmp/pip-build-env-vpmi0jy6/overlay/lib/python3.13/site-packages/setuptools/build_meta.py", line 333, in get_requires_for_build_wheel
          return self._get_build_requires(config_settings, requirements=[])
                 ~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "/tmp/pip-build-env-vpmi0jy6/overlay/lib/python3.13/site-packages/setuptools/build_meta.py", line 303, in _get_build_requires
          self.run_setup()
          ~~~~~~~~~~~~~~^^
        File "/tmp/pip-build-env-vpmi0jy6/overlay/lib/python3.13/site-packages/setuptools/build_meta.py", line 521, in run_setup
          super().run_setup(setup_script=setup_script)
          ~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "/tmp/pip-build-env-vpmi0jy6/overlay/lib/python3.13/site-packages/setuptools/build_meta.py", line 319, in run_setup
          exec(code, locals())
          ~~~~^^^^^^^^^^^^^^^^
        File "<string>", line 51, in <module>
        File "<string>", line 48, in _guard_py_ver
      RuntimeError: Cannot install on Python version 3.13.0; only versions >=3.9,<3.13 are supported.
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
error: subprocess-exited-with-error

× Getting requirements to build wheel did not run successfully.
│ exit code: 1
╰─> See above for output.

note: This error originates from a subprocess, and is likely not a problem with pip.


### Expected outcome

Should install Librosa

### Additional information

_No response_

### Operating system

Fedora 41 

### software versions

- prinLinux-6.11.4-201.fc40.x86_64-x86_64-with-glibc2.40
- Python 3.13.0 (main, Oct  8 2024, 00:00:00) [GCC 14.2.1 20240912 (Red Hat 14.2.1-3)]
- NumPy 2.1.2
- SciPy 1.14.1
- Librosa N.A.

### Installation

pip

**Comment by bmcfee** [comment]
There's nothing we can do about this until numba gets an update.  However, we don't currently claim to support 3.13, so I don't think it's accurate to call this a bug:
https://github.com/librosa/librosa/blob/d5aa7e1ae443c917fa53bef31055ed2613d63133/setup.cfg#L46-L60

I'll leave this open so we can keep tabs on it and update when numba does though.

**Comment by cdbruyla** [comment]
Hi @bmcfee, absolutely fair response, I did not check if you were supposed to support 3.13 by now :upside_down_face: 

Fair enough - thanks!

**Comment by SonuKumar145** [comment]
thanks for reporting the issue. i was so confused why it was not getting installed

**Comment by remysedlak** [comment]
one day...

**Comment by bmcfee** [comment]
I believe this is now unstuck with the release of numba 0.61 a few days ago.  Once the new numba packages hit conda-forge, we can add it to the CI matrix to ensure proper testing.  Assuming nothing else causes trouble, we should be able to add official support to the 0.11 release.

**Comment by bmcfee** [comment]
I just spun up a local test environment to try this out (using numba from pypi and everything else from conda-forge).

The only other problem that I'm seeing is that audioread does not work out of the box on 3.13 due to purged deprecations in the standard library (aifc and sunau).  This is documented here: https://github.com/beetbox/audioread/issues/144   There's a PR open on audioread for this, but it seems like there are some non-obvious issues to contend with before it will be merged and released.

Meanwhile, these missing packages can be shimmed around by installing `standard-aifc` and `standard-sunau` packages.  Having done this in my test environment, the test suite passes with no modifications.  So that's good news.

To summarize, we're blocked by:

- [ ] numba 0.61 on conda-forge  (this is an easy one)
- [ ] audioread 3.0.2 (?) with 3.13 support

Alternatively, we could take this opportunity to downgrade audioread to an optional dependency.  It's already deprecated since 0.10, so a downgrade to optional status in 0.11 seems reasonable before it is removed entirely in 1.0.  

There are some technical hiccups to making audioread optional: we'd have to do a little bit of a dance around imports to make the type annotation for `load` work, and otherwise restructure some of the fallbacks for `get_duration` and `get_samplerate`.  I think it's doable though.

**Comment by bmcfee** [comment]
I gave this some thought over the weekend, and I think it's best to just leave the code on our side as is, and punt the dependency constraints up to audioread.

Given that, I think we can add this (3.13 support) to the 0.11 docket since both of our upstream blockers should be resolved relatively quickly.

**Comment by bmcfee** [comment]
Quick update on this: numba 0.61 works on pypi, but not on conda-forge.  Since our CI is based on conda-forge, testing is blocked until numba packages get fixed.

Still waiting on a new audioread release to drop, but everything should otherwise work out of the box with a pip install.

My plan now is to cut a 0.11.0.rc0 pre-release to see how things shake out.  We'll have to wait for dependencies to properly resolve before a final release can be made, but I don't mind have a pre-release out that is fine except for a few obscure cases (eg loading .au files on python 3.13).

**Comment by uis246** [comment]
Conda-forge should have python3.13 version now

**Comment by bmcfee** [comment]
0.11.0 release is making its way out now.  Our upstream problems are not resolved, but I've set up the requirements to patch around the problems for now.

I'll close this out, but please chime in if the 0.11.0 version does not work for you.

**Comment by mishushakov** [comment]
I couldn't install v0.11.0 on 3.13.3 due to the numba error:

```
726.0 Collecting numba>=0.51.0 (from librosa==0.11.0->-r requirements.txt (line 21))
726.0   Downloading numba-0.61.0-cp313-cp313-manylinux_2_28_aarch64.whl.metadata (2.8 kB)
726.1   Downloading numba-0.60.0.tar.gz (2.7 MB)
727.0      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.7/2.7 MB 3.2 MB/s eta 0:00:00
727.1   Installing build dependencies: started
728.6   Installing build dependencies: finished with status 'done'
728.6   Getting requirements to build wheel: started
728.6   Getting requirements to build wheel: finished with status 'error'
728.6   error: subprocess-exited-with-error
728.6   
728.6   × Getting requirements to build wheel did not run successfully.
728.6   │ exit code: 1
728.6   ╰─> [24 lines of output]
728.6       Traceback (most recent call last):
728.6         File "/usr/local/lib/python3.13/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 389, in <module>
728.6           main()
728.6           ~~~~^^
728.6         File "/usr/local/lib/python3.13/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 373, in main
728.6           json_out["return_val"] = hook(**hook_input["kwargs"])
728.6                                    ~~~~^^^^^^^^^^^^^^^^^^^^^^^^
728.6         File "/usr/local/lib/python3.13/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 143, in get_requires_for_build_wheel
728.6           return hook(config_settings)
728.6         File "/tmp/pip-build-env-j88zk3wq/overlay/lib/python3.13/site-packages/setuptools/build_meta.py", line 331, in get_requires_for_build_wheel
728.6           return self._get_build_requires(config_settings, requirements=[])
728.6                  ~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
728.6         File "/tmp/pip-build-env-j88zk3wq/overlay/lib/python3.13/site-packages/setuptools/build_meta.py", line 301, in _get_build_requires
728.6           self.run_setup()
728.6           ~~~~~~~~~~~~~~^^
728.6         File "/tmp/pip-build-env-j88zk3wq/overlay/lib/python3.13/site-packages/setuptools/build_meta.py", line 512, in run_setup
728.6           super().run_setup(setup_script=setup_script)
728.6           ~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^
728.6         File "/tmp/pip-build-env-j88zk3wq/overlay/lib/python3.13/site-packages/setuptools/build_meta.py", line 317, in run_setup
728.6           exec(code, locals())
728.6           ~~~~^^^^^^^^^^^^^^^^
728.6         File "<string>", line 51, in <module>
728.6         File "<string>", line 48, in _guard_py_ver
728.6       RuntimeError: Cannot install on Python version 3.13.3; only versions >=3.9,<3.13 are supported.
728.6       [end of output]
728.6   
728.6   note: This error originates from a subprocess, and is likely not a problem with pip.
728.6 error: subprocess-exited-with-error
728.6 
728.6 × Getting requirements to build wheel did not run successfully.
728.6 │ exit code: 1
728.6 ╰─> See above for output.
728.6 
728.6 note: This error originates from a subprocess, and is likely not a problem with pip.
```

**Comment by bmcfee** [comment]
> 726.0 Collecting numba>=0.51.0 (from librosa==0.11.0->-r requirements.txt (line 21))
> 726.0   Downloading numba-0.61.0-cp313-cp313-manylinux_2_28_aarch64.whl.metadata (2.8 kB)
> 726.1   Downloading numba-0.60.0.tar.gz (2.7 MB)
> 727.0      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.7/2.7 MB 3.2 MB/s eta 0:00:00

This looks like it's trying to install numba 0.60.0, which did not support python 3.13.  Librosa would not force this through its dependency spec, but it looks like you have a separate requirements.txt that you're using here which might.

**Comment by mishushakov** [comment]
I do not have numba specified on my requirements.txt, maybe it's another package that is requesting a specific version? But I am not seeing any errors related to version conflicts

**Comment by bmcfee** [comment]
> I do not have numba specified on my requirements.txt, maybe it's another package that is requesting a specific version?

Can you post the requirements.txt here?

**Comment by mishushakov** [comment]
i have fixed it by specifying a later version of numba in my requirements.txt

thanks for the help tough!
